text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,558 @@
+  export type InvalidUnion = ListOfFoo | Bar;
+  ```
+
+In order to preserve backwards-compatbility, and to retain ability to use this new feature against the current CDK v1
+codebase, the `jsii` compiler will not **require** that _type unions_ are systematically aliased. Similarly,
+`jsii-pacmak` will be modified in such a way that it continues producing the same code whether a _type union_ is aliased
+or not.
+
+Once this done, the `cfn2ts` tool, which is used in the CDK to generate _CloudFormation Resources_ (also known as _L1
+constructs_) will be modified so that it emits aliases for any _type union_ it generates.
+
+Gearing up for the [AWS CDK v2] release, `jsii` will be updated to accept a new option (`--strict=type-unions`) that
+will allow users to opt into **requiring** that any visible _type union_ is aliased. This feature will be gradually
+turned on in the AWS CDK v1 development branch, as all existing _type unions_ are aliased. Once the whole codebase has
+been updated so that no _type union_ is unaliased, it will be ready for the introduction of **breaking changes** as part
+of [AWS CDK v2].
+
+## Breaking changes
+
+In order to deliver the full potential of this solution, code generation needs to be changed in `jsii-pacmak` so that
+languages without native support for _type unions_ (**C#**, **Go**, and **Java**) have additional _union-like_ types
+emitted.
+
+In order to avoid having to cut a new major version of `jsii-pacamk`, the new code generation patterns will be gated by
+an opt-in flag (`--union-like-types`). When the flag is enabled, aliased _union types_ have a known name that can be
+used to generate the _union-like_ types. The proposed API for a union such as
+`export type UnionType = Foo | Map<Bar> | Baz[]` is the following:
+
+- In **C#**
+
+  ```csharp
+  // UnionType.cs
+  public sealed class UnionType {
+    public static UnionType FromFoo(Foo value) { /* ... */ }
+    public static UnionType FromMapOfBar(IReadOnlyDictionary<Bar> value) { /* ... */ }
+    public static UnionType FromListOfBaz(IReadOnlyList<Baz> value) { /* ... */ }
+
+    public Foo? Foo { get; }
+    public IReadOnlyDictionary<Bar>? MapOfBar { get; }
+    public IReadOnlyList<Baz>? ListOfBaz { get; }
+  }
+  ```
+
+- In **Go**:
+
+  > :construction: TBD
+
+- In **Java**:
+
+  ```java
+  // UnionType.java
+  public final class UnionType {
+    public static UnionType fromFoo(Foo value) { /* ... */ }
+    public static UnionType fromMapOfBar(Map<String, Bar> value) { /* ... */ }
+    public static UnionType fromListOfBaz(Baz... value) { /* ... */ }
+    public static UnionType fromListOfBaz(List<Baz> value) { /* ... */ }
+
+    public @Nullable Foo asFoo() { /* ... */ }
+    public @Nullable Map<String, Bar> asMapOfBar() { /* ... */ }
+    public @Nullable List<Baz> asListOfBaz() { /* ... */ }
+  }
+  ```
+
+The _jsii Kernel_ will need adjustments in order to better support those types. The generated libraries will provide a
```

Comment:
Can we do this without a flag?

Generally, I am wondering - since today Java/.NET code just uses `Object` for union types, we can technically move to generate the union-like classes and accept them as inputs for those `Object` properties and then, the only thing we do in v2.x is replace `Object` with the union name. 

The benefit is that we basically already generate all the classes and allow them to be used freely and only introduce the breaking API change in 2.x.

Reply:
Except this consists in a breaking change for Java customers. Right now, they coud get the `Object` and in many cases down-cast that to the correct dynamic type / check it with an `instanceof` guard.

If we change the dynamic type of return values of `Object`-returning methods, then all of these will suddenly break.